dart/dynamics/Frame.h

```diff
@@ -0,0 +1,297 @@
+ *   AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ *   LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
+ *   ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ *   POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#ifndef DART_DYNAMICS_FRAME_H_
+#define DART_DYNAMICS_FRAME_H_
+
+#include <set>
+
+#include <Eigen/Geometry>
+
+#include "dart/dynamics/Entity.h"
+#include "dart/math/MathTypes.h"
+
+namespace dart {
+namespace dynamics {
+
+/// The Frame class serves as the backbone of DART's kinematic tree structure.
+///
+/// Frame inherits Entity, so it exists within a reference Frame. This class
+/// keeps track of both its local (relative) and global (world) transforms,
+/// velocities, and accelerations. It also notifies every child Entity when
+class Frame : public virtual Entity
+{
+public:
+  friend class Entity;
+  friend class WorldFrame;
+
+  /// Constructor for typical usage
+  explicit Frame(Frame* _refFrame, const std::string& _name);
+
+  /// Destructor
+  virtual ~Frame();
+
+  static Frame* World();
+
+  //--------------------------------------------------------------------------
+  // Transform
+  //--------------------------------------------------------------------------
+
+  /// Get the transform of this Frame with respect to its parent Frame
+  virtual const Eigen::Isometry3d& getRelativeTransform() const = 0;
+
+  /// Get the transform of this Frame with respect to the World Frame
+  const Eigen::Isometry3d& getWorldTransform() const;
+
+  /// Get the transform of this Frame with respect to some other Frame
+  Eigen::Isometry3d getTransform(
+      const Frame* _withRespectTo = Frame::World()) const;
+
+  //-------------------------------------------------------------------------
+  // Velocity
+  //-------------------------------------------------------------------------
+
+  /// Get the spatial velocity of this Frame relative to its parent Frame, in
+  /// its own coordinates.
+  virtual const Eigen::Vector6d& getRelativeSpatialVelocity() const = 0;
+
+  /// Get the total spatial velocity of this Frame in the coordinates of this
+  /// Frame.
+  const Eigen::Vector6d& getSpatialVelocity() const;
+
+  /// Get the total spatial velocity of this Frame. The velocity can be provided
+  /// in the coordinates of any Frame.
+  Eigen::Vector6d getSpatialVelocity(const Frame* _inCoordinatesOf) const;
+
+  /// Get the spatial velocity of this Frame relative to some other Frame. It
+  /// can be expressed in the coordinates of any Frame.
+  Eigen::Vector6d getSpatialVelocity(const Frame* _relativeTo,
+                                     const Frame* _inCoordinatesOf) const;
+
+  /// Get the linear portion of classical velocity of this Frame relative to
+  /// some other Frame. It can be expressed in the coordinates of any Frame.
+  Eigen::Vector3d getLinearVelocity(
+      const Frame* _relativeTo = Frame::World(),
+      const Frame* _inCoordinatesOf = Frame::World()) const;
```

Comment:
I was actually planning on offering this functionality with the Point class that I've mentioned in the past. It would look something like this:

Eigen::Vector3d linearVelocity = Point(Eigen::Vector3d(1,0,0), referenceFrame).getLinearVelocity();

I could also change the functions to work the way you've proposed, but there are two reasons I didn't: (1) The number of arguments starts to look kind of crazy, and (2) If we're always accounting for an offset, that means there are some extra computations that we are always doing.

I suppose both of those problems can be solved by offering two different versions of the function: one with an offset argument (it would be the first argument in the function) and one without an offset argument. Does that seem like a decent compromise?